### DISCUSSION TITLE
Is there a way in 3.x to specify multiple features / bundles repos for pom-less builds

### DISCUSSION THREAD

**Discussion by peterkir** [eclipse-tycho/tycho#2363] [Q&A] [answered]
Hi all,

I have a question regarding tycho 3.x and structured builds.
As I was not able to find documentation about it, I'm asking here.
In a complex system, that has for organizational purposes different folders for specific domains containing feature / bundle sets.

<root>
<root>pom.xml                            # parent.pom
<root>/feature_domain_A
<root>/plugins_domain_A
<root>/feature_domain_B
<root>/plugins_domain_B
<root>/releng ....

Is this feasible and how can it be configured it it is?

**Comment by laeubi** [comment]
> As I was not able to find documentation about it, I'm asking here.

The documentation is here please suggest changes if anything is missing:
https://tycho.eclipseprojects.io/doc/master/StructuredBuild.html

> Is this feasible and how can it be configured it it is?

You can can reference any bundle or feature directly in the `<modules>` section, beside that, you can reference any folder containing bundles / features in the  `<modules>` section if it is named `plugins`, `bundles` or `features` and all items will automatically be aggregated.

**Reply by peterkir** [reply]
Thx for the doc link, 
I tried to specify sub domain modules with
`<module>domain1/features</module>
<module>domain1/bundles</module>`
and with
`<module>domain1`</module>`
given a "parent-pom"
```
<project>
  <modelVersion>4.0.0</modelVersion>
  <parent>
    <groupId>example.rcp</groupId>
    <artifactId>example.rcp.parent</artifactId>
    <version>1.0.0-SNAPSHOT</version>
    <relativePath>..</relativePath>
  </parent>
  <groupId>example.rcp</groupId>
  <artifactId>example.rcp.domain1</artifactId>
  <version>1.0.0-SNAPSHOT</version>
  <packaging>pom</packaging>
  <modules>
    <module>bundles</module>
  </modules>
</project>
```
Sadly without luck.
Can you give me hint how to specify the additional `domain1` feature / bundles?
I have a [sample repo](https://github.com/klibio/example.pde.rcp) where I try to evaluate.

**Reply by laeubi** [reply]
> Sadly without luck.

In any case its good to explain what you _expected_ and what you get _instead_.

> I have a [sample repo](https://github.com/klibio/example.pde.rcp) where I try to evaluate.

It complains that it can't find a "root" pom in your `domain1` folder, so you can place a pom there or configure the path to the parent:
https://github.com/eclipse-tycho/tycho/wiki/Tycho-Pomless#specify-parent-path

**Reply by peterkir** [reply]
I added the file structure and my current understanding into the README.
currently I get the error `Project 'example.rcp:bundles:1.0.0-SNAPSHOT' is duplicated in the reactor`
So my question are
- what to specifiy in the top level parent `pom.xml` for the `domain1` sub-folder
- is a "connector" `domain1/pom.xml` required?
- is the `domain1/bundles-b1/build.properties` specified `tycho.pomless.parent = ../../..` and correct?

**Reply by laeubi** [reply, marked as answer]
You can place a `build properties` file in the `domain1/bundles` folder and there configure the parent and a different module name (so you don't need this in each individual bundle!):
https://github.com/eclipse-tycho/tycho/wiki/Tycho-Pomless#overwrite-group-and-artifact-ids

**Reply by peterkir** [reply]
Thx for help. Finallly I got. 
I updated the [wiki page](https://github.com/eclipse-tycho/tycho/wiki/Tycho-Pomless) to make it more detailed for slow guys like myself :-)  Hope that is fine with you.

**Comment by dhendriks** [comment]
In my experience, if you have a pom-less Tycho build, and you use non-default folder names in the root of your repo, that contain your plugins/features/etc, you need to run Maven with `-Dtycho.pomless.aggregator.names=folder1,folder2,folder3` to make Maven/Tycho aware of this. Just adding them as `modules` to `pom.xml` is not enough. See also [this example](https://gitlab.eclipse.org/eclipse/escet/escet/-/blob/3682256d0fc3969d0fcfafbb84c362fa196415df/mvn_escet.sh#L17).

**Reply by laeubi** [reply]
@dhendriks this is only required if you use non "standard" names, what is not the case here.

**Reply by dhendriks** [reply]
The `different folders for specific domains containing feature / bundle sets` part of the original question made me think that indeed non-standard names are being used.
